### PR TITLE
feat!: new icon approach

### PR DESCRIPTION
**Pull request by antfu**
This PR introduces a brand new icon approach that works much better with SSR (no flicking on the client side) and should help reduce the load of Iconify's public API by a lot.

### Breaking Changes

- `<Icon>` and `<IconCSS>` is merged together, use `<Icon mode="svg">` and `<Icon mode="css">`
- `<Icon>` now default using `css` mode, (set `icon.defaultMode: 'svg'` to use the previous beahvor
- `<IconCSS>` component is removed
- Emoji-as-name is no longer supported
- Slot fallback is not supported yet
- App config key changed from `nuxtIcon` to `icon`
- App config schema changes
  - `iconifyApiOptions` has been removed
  - `iconifyApiOptions.url` -> `iconifyApiEndpoint`
  - `iconifyApiOptions.publicApiFallback` -> `fallbackToApi` (default: true)

### New Features

- Support loading icons for local icons folder, close https://github.com/nuxt-modules/icon/issues/131

### Approaches

- Create a `/api/_nuxt_icon/` endpoint that will serve the icons from server-bundle, or fallback to Iconify API (fallback to use public API when `ssr:false`
- **CSS Mode**: `<Icon name="icon-name" `mode="`CSS`" /`>` will render the icon as a CSS class, SSR is supported by injecting compiled CSS into HTML.
  - On the client side, it will look up the registered CSS selectors. If the class is already presented, it will skip icon rendering (compatible with pre-compiled UnoCSS icons-as-css solution, TW support is still in research)
- **SVG Mode**: `<Icon name="icon-name" />` component will use `useAsyncState` to fetch the icon data from the server (local fetch when SSR) and force Iconify to render the icon to SSR. Hydrate the data back on client side.
  - For dynamic icons that are presented on the server side, no client request will be made.
  - For dynamic icons that are presented only on the client side, a client request to `/api/_nuxt_icon/` will be made to fetch the icon data (cached event handler).

### TODOs

- [x] Update docs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
   "name": "nuxt-icon",
+  "packageManager": "pnpm@9.0.5",
   "version": "0.6.10",
   "license": "MIT",
   "type": "module",
   "repository": "nuxt-modules/icon",
   "exports": {
     ".": {
-      "types": "./dist/types.d.ts",
+      "types": "./dist/module.d.ts",
```
Comment (benjamincanac) [inline package.json:10]: @antfu Why not use `./dist/types.d.ts` anymore?

File: src/module.ts [hunk 2]
```diff
@@ -21,102 +17,124 @@ export default defineNuxtModule<ModuleOptions>({
...
+    const serverKnownCssClasses = options.serverKnownCssClasses || []
+    nuxt.options.runtimeConfig.nuxtIcon = {
+      serverKnownCssClasses,
+    }
+    nuxt.hook('nitro:init', async (_nitro) => {
+      _nitro.options.runtimeConfig.nuxtIcon = {
+        serverKnownCssClasses,
+      }
+    })
+    unocssIntegration(nuxt, options)
```
Comment (benjamincanac) [inline src/module.ts:128]: Is it necessary to call `unocssIntegration` here even though it might not be installed?

File: package.json (hunk 1 above)
Comment (antfu) [reply package.json:10]: I am not sure, but the latest module-builder no longer generates that file for some reason.

File: src/module.ts [hunk 3]
```diff
@@ -21,102 +18,131 @@ export default defineNuxtModule<ModuleOptions>({
...
+    nuxt.options.runtimeConfig.icon = {
+      serverKnownCssClasses,
+    }
+    nuxt.hook('nitro:init', async (_nitro) => {
+      _nitro.options.runtimeConfig.icon = {
+        serverKnownCssClasses,
+      }
+    })
+
+    if (hasModule(nuxt, '@unocss/nuxt'))
```
Comment (benjamincanac) [inline src/module.ts:130]: Why not use `hasNuxtModule` from `@nuxt/kit`?

Comment (antfu) [reply src/module.ts:130]: Haha cool, I didn't know it exists!

**Comment by benjamincanac** [comment]
@antfu Would that be pertinent to add an option to append `aria-hidden="true"` on every icon?

**Comment by antfu** [comment]
Maybe a `defaultAttrs` option in the app config, what do you think?

**Comment by benjamincanac** [comment]
I'm not sure it's a good idea for `@nuxt/ui` to configure this for all icons but this could be a nice addition anyway for `nuxt-icon` users 😊

**Comment by antfu** [comment]
Alright, I think it should be good to go for a few beta releases.

**Comment by gitFoxCode** [comment]
>(set icon.defaultMode: 'svg' to use the previous beahvor)

its `icon.defaultMode: 'svg'` or `icon.mode: 'svg'` 😅

**Pull request merged**